File: app/src/main/java/com/donald/musictheoryapp/Question/MultipleChoiceQuestion.java
package com.donald.musictheoryapp.Question;

import com.donald.musictheoryapp.Utils.JSONArrayUtil;

import org.json.JSONException;
import org.json.JSONObject;

public class MultipleChoiceQuestion extends Question
{
    // TODO: USE INTS
    public enum OptionType
    {
        TEXT,
        IMAGE,
        SCORE
    }

    public static class Answer implements Question.Answer
    {
        public int userAnswer;
        public int correctAnswer;

        public Answer(int userAnswer, int correctAnswer)
        {
            this.userAnswer = -1;
            this.correctAnswer = correctAnswer;
        }

        @Override
        public boolean correct()
        {
            return userAnswer == correctAnswer;
        }

        public static Answer fromJSON(JSONObject object) throws JSONException
        {
            return new Answer(
                object.isNull("user_answer") ? -1 : object.getInt("user_answer"),
                object.getInt("correct_answer")
            );
        }
    }

    public Answer answer;
    public String[] options;
    public OptionType optionType;

    public MultipleChoiceQuestion() {}

    @Override
    public int points()
    {
        if(answer.correct()) return 1;
        return 0;
    }

    @Deprecated
    public String[] getOptions() { return options; }

    @Deprecated
    public OptionType getOptionType() { return optionType; }

    public void acceptVisitor(QuestionVisitor visitor) { visitor.visit(this); }

    public static MultipleChoiceQuestion fromJSON(JSONObject object, QuestionGroup group) throws JSONException
    {
        MultipleChoiceQuestion question = new MultipleChoiceQuestion();
        question.number = object.getInt("number");
        question.group = group;
        question.descriptions = JSONArrayUtil.descriptions(object);
        question.optionType = OptionType.values()[object.getInt("option_type")];
        question.options = JSONArrayUtil.options(object);
        question.answer = Answer.fromJSON(object.getJSONObject("answer"));
        return question;
    }
}

File: app/src/main/java/com/donald/musictheoryapp/question/MultipleChoiceQuestion.kt
package com.donald.musictheoryapp.question

import com.donald.musictheoryapp.util.*
import kotlin.Throws
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

class MultipleChoiceQuestion(
    number: Int,
    descriptions: List<Description>,
    inputHint: String?,
    val options: List<String>,
    val optionType: OptionType,
    val answer: Answer
) : ChildQuestion(number, descriptions, inputHint) {

    override val points: Int
        get() = if (answer.correct) 1 else 0
    override val maxPoints = 1
    override val correctAnswerCount = answer.correctAnswers.size

    override fun acceptVisitor(visitor: QuestionVisitor) {
        visitor.visit(this)
    }

    override fun getAdditionalQuestionImagesRequired(): List<String> {
        return if (optionType == OptionType.Image) {
            options
        } else {
            emptyList()
        }
    }

    override fun registerImages(imageArray: JSONArray) {
        super.registerImages(imageArray)
        if (optionType == OptionType.Image) {
            for (option in options) {
                imageArray.put(option)
            }
        }
    }

    override fun copyNew(): MultipleChoiceQuestion {
        return MultipleChoiceQuestion(
            number,
            List(descriptions.size) { i -> descriptions[i].copy() },
            inputHint,
            List(options.size) { i -> options[i] },
            optionType,
            answer.copyNew()
        )
    }

    override fun toPartialJson(): JSONObject {
        val jsonObject = JSONObject()
        val array = JSONArray()
        jsonObject.put(QUESTION_TYPE, Type.MULTIPLE_CHOICE.ordinal)
        for (option in options) array.put(option)
        jsonObject.put("options", array)
        jsonObject.put(OPTION_TYPE, optionType.ordinal)
        jsonObject.put("answer", answer.toJson())
        return jsonObject
    }

    companion object {

        @Throws(JSONException::class)
        fun fromJson(jsonObject: JSONObject): MultipleChoiceQuestion {
            return MultipleChoiceQuestion(
                number = jsonObject.getInt("number"),
                descriptions = jsonObject.getDescriptions(),
                inputHint = jsonObject.getInputHint(),
                options = jsonObject.getOptions(),
                optionType = jsonObject.getOptionType(),
                answer = Answer.fromJson(jsonObject.getJSONObject("answer")),
            )
        }

    }

    enum class OptionType { Text, Image, Score }

    class Answer(var userAnswer: Int?, val correctAnswers: List<Int>) : ChildQuestion.Answer {

        override val correct: Boolean
            get() = correctAnswers.any { it == userAnswer }

        override fun copyNew(): Answer {
            return Answer(
                null,
                List(correctAnswers.size) { i -> correctAnswers[i] }
            )
        }

        fun toJson(): JSONObject {
            val jsonObject = JSONObject()
            jsonObject.put("user_answer", userAnswer ?: JSONObject.NULL)
            jsonObject.put(
                "correct_answers",
                JSONArray().apply { correctAnswers.forEach { put(it) } }
            )
            return jsonObject
        }

        companion object {

            fun fromJson(jsonObject: JSONObject): Answer {
                return Answer(
                    if (jsonObject.isNull("user_answer")) null else jsonObject.getInt("user_answer"),
                    jsonObject.getCorrectAnswers()
                )
            }

        }

    }

}